src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,180 @@
 """Implementation of Rule L003."""
-from typing import List, Optional, Sequence, Tuple
+import dataclasses
+import itertools
+from typing import Dict, List, Optional, Sequence, Tuple
 
 from sqlfluff.core.parser import WhitespaceSegment
-from sqlfluff.core.parser.segments import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments import BaseSegment
 from sqlfluff.core.rules.functional import rsp, sp, Segments
 from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
 from sqlfluff.core.rules.doc_decorators import (
     document_fix_compatible,
     document_configuration,
 )
 from sqlfluff.core.templaters import TemplatedFile
+from sqlfluff.core.templaters.base import RawFileSlice
+
+
+@dataclasses.dataclass
+class _LineSummary:
+    """A dataobject to represent a line.
+
+    Doubles as a memory object (holding state)
+    that can be converted to a fresh summary instance
+    """
+
+    line_no: int = 0
+    templated_line: Optional[int] = None
+    line_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_size: int = 1
+
+    # The final balance of a line once fully considered
+    indent_balance: int = 0
+    # The indent as it was once we say this lines "Anchor"
+    anchor_indent_balance: int = 0
+    hanging_indent: Optional[int] = None
```

Comment:
It could be helpful to reorder and group these fields a bit differently, by category, e.g.:
* Move `line_anchor` adjacent to `anchor_indent_balance`
* These in a separate section: `templated_line_type`, `is_comment_line`, `is_empty_line`

Not sure if these are exactly right, just thinking to group the fields by functional area.

Reply:
Fixedish ?